File: LEETCODE/ARRAY/74. Search a 2D Matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        # using binary search in 2d matrix
        if len(matrix) == 0: return False
            
        row = len(matrix)
        column = len(matrix[0])
        
        low =0
        high = (row*column)-1
        
        while low<=high:
            mid = (low+high)//2
            if matrix[mid//column][mid%column]==target:
                return True
            elif target < matrix[mid//column][mid%column]:
                high = mid-1
            else:
                low=mid+1
        return False
        
        
        
        
        
        
        
        #using brute force approach
#         for i in range(len(matrix)):
#             for j in range(len(matrix[i])):
#                 if matrix[i][j] == target:
#                     return True
#         else:
#             return False
        
        
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        

File: LEETCODE/ARRAY/74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0){
            return false;
        }
        int row = matrix.length;
        int column = matrix[0].length;
        
        int low = 0;
        int high = (row*column)-1;
        
        boolean flag = false;
        while (low <= high){
            int mid = (low+high)/2;
            if(matrix[mid/column][mid%column]==target){
                return true;
            }
            else if (target < matrix[mid/column][mid%column]){
                high = mid-1;
            }
            else{
                low = mid+1;
            }
        }
        return flag;
        
    }
}

